src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/listeners/MigrationMessageListener.kt

```diff
@@ -22,16 +25,20 @@ class MigrationMessageListener(
   @JmsListener(destination = "migration", containerFactory = "hmppsQueueContainerFactoryProxy")
   fun onMessage(message: String) {
     log.debug("Received message {}", message)
-    val migrationMessage: MigrationMessage = message.fromJson()
+    val migrationMessage: MigrationMessage<*> = message.fromJson()
     when (migrationMessage.type) {
-      MIGRATE_VISITS -> visitsMigrationService.migrateVisitsByPage()
+      MIGRATE_VISITS -> visitsMigrationService.migrateVisitsByPage(context(message.fromJson()))
     }
   }
 
-  data class MigrationMessage(
+  class MigrationMessage<T>(
     val type: Messages,
-    val body: Any? = null
+    val context: MigrationContext<T>
   )
 
-  private inline fun <reified T> String.fromJson(): T = objectMapper.readValue(this, T::class.java)
+  private inline fun <reified T> String.fromJson(): T =
+    objectMapper.readValue(this, object : TypeReference<T>() {})
```

Comment:
I could be wrong but I found that if you use a "Kotlin" objectMapper there is a clever single argument readValue() method.
i.e. use  objectMapper = ObjectMapper().registerModule(KotlinModule.Builder().build())

Reply:
cool, I'll looked at that